### PR TITLE
Add `AsyncInstance` class

### PR DESCRIPTION
**Pull request by irregularunit**
# About
As of [async_init](https://github.com/irregularunit/duck-hideout-manager-bot/blob/main/cogs/stats/calendar.py#L41) you have defined as an async coroutine with patching it with `async_init`.

This PR introduces a new `AsyncInstance` class that you can use to create async classes. This removes the need for you to patch the `__init__` method yourself, and hopefully making your code more readable and easier to maintain.

# Note
The current implementation needs better typings. I'm aware that the current implementation of the class does not strict type checking.

I find it too much of a hassle to fix the typing myself. That is why I cannot fix the annotations.

### CODE REVIEW

**Review by LeoCx1000 — CHANGES_REQUESTED** [review]
Hello. Thanks for your PR!

I would like to note a few things, first and foremost you cannot name a file `async`, since that keyword is reserved for defining async functions in python and it will fail to import.

Second, what would be the use of this class. That is, how should it be used? Adding a doc-string to `AsyncInstance` would be preferred so other contributors know how to use said class without having to read the code within it.

**Comment by irregularunit** [comment]
Hi there,

I've made some changes to this code:

> I would like to note a few things, first and foremost you cannot name a file `async`, since that keyword is reserved for defining async functions in python and it will fail to import.

- Sorry for the rookie mistake.

> Second, what would be the use of this class. That is, how should it be used? Adding a doc-string to `AsyncInstance` would be preferred so other contributors know how to use said class without having to read the code within it.

- Added a class-level docstring to provide clarity on how to use the class which hopefully clears any questions on how to use this class.

Sorry for the rookie mistake in the initial commit. Let me know if there are any further changes you'd like me to make.

Thanks.

**Comment by LeoCx1000** [comment]
Thanks for the clarification. After trying out your example, pyright does not recognise this class is awaitable.

Unless this is addressed, I can't push this PR forwards, I would rather not have to litter the code with `type: ignore` comments that can be avoided.

Code_KZh3W6bdOY

**Comment by irregularunit** [comment]
Hi there,

could you please review my pr again.

File: utils/async_utils.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
+from __future__ import annotations
+
+import asyncio
+import logging
+from abc import ABCMeta
+from contextlib import suppress
+from typing import Any, Callable, Coroutine, Generator, MutableSet, Optional, overload
```
Comment (LeoCx1000) [inline utils/async_utils.py]:
```suggestion
from typing import Any, Callable, Coroutine, Generator, MutableSet, Optional, overload, Self
```

File: utils/async_utils.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+    def compose_task(self, *args: Any, **kwargs: Any) -> asyncio.Task[Any]:
+        return self.__tasks__.compose_task(*args, **kwargs)
+
+    def compose_future(self) -> asyncio.Future[Any]:
+        return self.__tasks__.compose_future()
+
+    async def __adel__(self) -> None:
+        pass
+
+    def __await__(self) -> Generator[Any, None, AsyncInstance]:
```
Comment (LeoCx1000) [inline utils/async_utils.py]:
```suggestion
    def __await__(self) -> Generator[Any, None, Self]:
```

Return ``typing.Self`` here so the linter knows which subclass it is when awaiting it.

```py
test = await Test()
reveal_type(test)  # type is AsyncInstance -> type is Test
```

**Review by LeoCx1000 — CHANGES_REQUESTED** [review]
Thanks for addressing these changes. 

Another problem that I have encountered is that the linter does not recognise the arguments that are given to ``__ainit__`` so the intellisense for the constructor is missing, though I don't think that can be addressed.

Other than that, it all seems to work as expected.

**Comment by irregularunit** [comment]
Good morning,

thanks for the improvements and suggestions, is there anything else you would like me to change?

**Review by LeoCx1000 — APPROVED** [review]
Seems fine.

**Comment by LeoCx1000** [comment]
Thanks

**Pull request merged**